proc_macros/src/lib.rs

```diff
@@ -0,0 +1,350 @@
+            collection: &'collection #collection_type_name<T>,
+            next_index: usize,
+        }
+
+        impl<'collection, T> #values_struct_name<'collection, T> {
+            pub fn new(collection: &'collection #collection_type_name<T>) -> Self {
+                Self {
+                    collection,
+                    next_index: 0,
+                }
+            }
+        }
+
+        impl<'collection, T> Iterator for #values_struct_name<'collection, T> {
+            type Item = &'collection T;
+
+            fn next(&mut self) -> Option<Self::Item> {
+                if self.next_index < self.collection.len() {
+                    let ret = Some(&self.collection.0[self.next_index]);
+                    self.next_index += 1;
+                    ret
+                } else {
+                    None
+                }
+            }
+        }
+    }
+}
+
+fn get_index_implementations(
+    name: &Ident,
+    collection_type_name: &Ident,
+) -> proc_macro2::TokenStream {
+    quote! {
+        impl<T> Index<#name> for #collection_type_name<T> {
+            type Output = T;
+
+            fn index(&self, index: #name) -> &Self::Output {
+                &self.0[index as usize]
+            }
+        }
+
+        impl<T> Index<usize> for #collection_type_name<T> {
+            type Output = T;
+
+            fn index(&self, index: usize) -> &Self::Output {
+                &self.0[index]
+            }
+        }
+    }
+}
+
+fn get_deref_implementation(
+    collection_type_name: &Ident,
+    variants: &[Ident],
+) -> proc_macro2::TokenStream {
+    let len = variants.len();
+    quote! {
+        impl<T> Deref for #collection_type_name<T> {
+            type Target = [T; #len];
+
+            fn deref(&self) -> &Self::Target {
+                &self.0
+            }
+        }
+    }
+}
+
+fn get_from_usize_implementation(name: &Ident, variants: &[Ident]) -> proc_macro2::TokenStream {
+    quote! {
+        impl From<usize> for #name {
+            fn from(value: usize) -> Self {
+                match value {
+                    #(value if value == Self::#variants as usize => Self::#variants),*,
+                    _ => unreachable!(),
+                }
+            }
+        }
+    }
+}
+
+fn get_all_variants_collection_definition(
+    name: &Ident,
+    collection_type_name: &Ident,
+    variants: &[Ident],
+    all_variants_collection_name: &Ident,
+) -> proc_macro2::TokenStream {
+    quote! {
+        pub static #all_variants_collection_name: #collection_type_name<#name> = {
+            use SupportedLanguage::*;
+            BySupportedLanguage([
+                #(#variants),*
+            ])
+        };
+    }
+}
+
+fn get_collection_generator_macro_definition(
+    collection_type_name: &Ident,
+    variants: &[Ident],
+) -> proc_macro2::TokenStream {
+    let macro_name = format_ident!("{}", collection_type_name.to_string().to_snake_case());
+    quote! {
+        #[macro_export]
```

Comment:
Ok so this is sort of funny, the way that I was able to get a "generated" eg `by_supported_language!()` macro (that needed to be basically implemented as a proc macro, but per https://github.com/helixbass/tree-sitter-grep/pull/52#discussion_r1259916094 I don't think you could easily just "generate and use a new proc macro"?) was to generate it as a _declarative_ macro which basically delegates to an existing "static" `by_fixed_map!()` proc macro but has the things specific to "this fixed map" (ie the variant names and collection type name eg `BySupportedLanguage`) "pre-curried" as arguments to it (by hardcoding them into the declarative macro body basically)

It's sort of funny that I believe those pre-curried arguments just get passed along from one to the other "as token trees", so they can eg reference non-existing identifiers

Reply:
Ha yes, that's kind of ridiculous, but cool that it works.